app/queries/get_change_offer_options.rb

```diff
@@ -0,0 +1,84 @@
+class GetChangeOfferOptions
+  attr_accessor :user, :for_provider, :recruitment_cycle_year
+
+  # application_choice.offered_course (either current or the original)
+  def initialize(user:, current_course:)
+    @user = user
+    @for_provider = current_course.accredited_provider || current_course.provider
+    @recruitment_cycle_year = current_course.recruitment_cycle_year
+  end
+
+  def make_decisions_courses
+    Course
+      .joins(training_provider_make_decisions)
+      .joins(ratifying_provider_make_decisions)
+      .where(combine_user_and_provider_permissions)
+  end
+
+  def offerable_courses
+    make_decisions_courses
+      .where(
+        open_on_apply: true,
+        recruitment_cycle_year: recruitment_cycle_year,
+      )
+      .where(ratifying_provider_is_preserved)
+  end
+
+  def available_providers
+    Provider
+      .with(offerable_courses: offerable_courses)
+      .joins('INNER JOIN offerable_courses ON providers.id = offerable_courses.provider_id')
+      .distinct
+  end
+
+private
+
+  def permitted_provider_ids
+    user.provider_permissions.where(make_decisions: true).pluck(:provider_id)
+  end
+
+  def provider_relationship_permissions_sql(join_name, additional_checks)
+    <<~PROVIDER_RELATIONSHIP_PERMISSIONS.squish
+      LEFT OUTER JOIN provider_relationship_permissions AS #{join_name}
+        ON provider_id = #{join_name}.training_provider_id
+        AND accredited_provider_id = #{join_name}.ratifying_provider_id
+        AND #{additional_checks}
+    PROVIDER_RELATIONSHIP_PERMISSIONS
+  end
+
+  def training_provider_make_decisions
+    check = 'training_provider_make_decisions.training_provider_can_make_decisions IS TRUE'
+    provider_relationship_permissions_sql('training_provider_make_decisions', check)
+  end
+
+  def ratifying_provider_make_decisions
+    check = 'ratifying_provider_make_decisions.ratifying_provider_can_make_decisions IS TRUE'
+    provider_relationship_permissions_sql('ratifying_provider_make_decisions', check)
+  end
+
+  def combine_user_and_provider_permissions
+    return '1 != 1' if permitted_provider_ids.blank?
```

Comment:
What does `1 != 1` mean? 
Can it be moved to a well named constant so this line is easy to understand.

Reply:
If your user has no providers, by definition you are not authorised to make decisions on anything. By adding `1 != 1` to the where clause we guarantee no courses are returned while still supplying valid SQL for the `where`. Without it, an exception is thrown because `provider_id IN ()` is not valid SQL.

Reply:
Added a comment